libpromises/expand.c

```diff
@@ -188,13 +188,15 @@ static void MapIteratorsFromRval(EvalContext *ctx,
 }
 
 static PromiseResult ExpandPromiseAndDo(EvalContext *ctx, PromiseIterator *iterctx,
-                                        PromiseActuator *act_on_promise, void *param)
+                                        PromiseActuator *act_on_promise, void *param,
+                                        bool do_ifelse)
 {
     PromiseResult result = PROMISE_RESULT_SKIPPED;
 
     /* TODO this loop could be completely skipped for for non vars/classes if
      *      act_on_promise is CommonEvalPromise(). */
-    while (PromiseIteratorNext(iterctx, ctx))
+    bool done = ! do_ifelse;
```

Comment:
This translation from `do_ifelse` to `done` is very confusing and deserves a big comment. Or maybe actually working with `do_ifelse` would make things clearer?